Add MDCR support for P2000t

pokowaka opened the pull request:
This adds support for the mini digital cassette recorder that can be
found inside a P2000t. This implementation is based on documentation
that can be found in https://github.com/p2000t/documentation.

I did some reverse engineering of the rom to discover:

- The RDC flip-flop is not well documented and just casually mentioned.
- There is no documentation that mentions the P2000 reads from tape when
the drive is moving in reverse.
- There is no mention that the RDC/RDA signals are swapped when
the tape is moving in reverse.

This emulation is not timing perfect, but within the bounds of what is
acceptable to the rom and diagnostics module.

cuavas reviewed (changes requested):
Please follow MAME formatting conventions:
* Indent using tabs, with one tab per scope level
* Indent initialiser lists by one level
* Indent statement continuations (e.g formal parameter lists too long for one line) by two levels to make it clear they aren’t a scope level
* Don’t move right-associative qualifiers (e.g. reference `&` and pointer `*`) to the left of the space, keep them with the token they associate with
* Follow the convention for `#include` guard macro names – they should match the path/filename
* Use `snake_case` for class names and public member function names
* Don’t use overly generic names for classes in the global namespace

Why have you created a bus and slot here?  Does the system actually have a slot that accepts interchangeable devices, populated with a tape drive by default, or is the tape drive effectively an integral part of the system?  The bus/slot system should only be used for interchangeable devices.

Why are you using mutexes?  Why are you using `std::bitset` when you don’t need any of the things it offers over an unsigned integer?

pokowaka commented:
Thank you for the quick review!

- I'm not too familiar with the Mame code base and looked at other devices (PET) to make it to work. The tape is an integral part of the machine, so it sounds like it needs a different abstraction. Please let me know which abstraction to use and I'll move the code around.

I will fix the other issues:

- I used a clang-format file which clearly doesn't do the right thing.  (btw, is there a clang-format file that I could use?)
- std::mutex were left over from earlier tests where access was not thread safe, I can remove those.
- std::bitset provides an easy to understand abstraction which should not result additional overhead. I will switch to uint as it sounds that thats the preferred approach.

Do you prefer additional commits addressing your concerns or do you want me to update this change?

rb6502 commented:
For things that are integral parts of the machine, just create a class inheriting from device_t and put it in src/mame/machine.  Additional updates to this PR to correct things are fine.

cuavas commented:
I’d turn it not a device in src/mame/machine to preserve the encapsulation.  For a somewhat trivial example of a media device that isn’t a slot device in MAME, src/mame/machine/imm6_76.h and src/mame/machine/imm6_76.cpp which implement an EPROM programmer used by src/mame/drivers/intellec4.cpp (the same device is also present on some other Intel development systems that aren’t currently emulated in MAME).

MAME runs the machine scheduler on a single thread, so you don’t need to use mutexes to serialise operations unless you’re implementing your own work queues, and if you’re doing that you also need to ensure you’re dealing with task dependencies properly.

pokowaka commented:
@cuavas, thank you for the pointers. I will refactor with your suggestions.

pokowaka commented:
I've added the following changes:

- Addressed CR feedback
- Add support for .cas files.

pokowaka commented:
Looks like the Visual Studio build fails due to unrelated reasons:

```
error: perl: signature from "David Macek <david.macek.0@gmail.com>" is invalid
:: File /var/cache/pacman/pkg/perl-5.32.0-2-x86_64.pkg.tar.zst is corrupted (invalid or corrupted package (PGP signature)).
```

cuavas commented:
Significant chunks of src/mame/includes/p2000t.h and src/mame/machine/p2000t.cpp have whitespace changes from previously being reformatted.  Can you please revert the spurious changes?

cuavas commented on src/mame/machine/p2000t_mdcr.cpp:
```diff
@@ -0,0 +1,323 @@
+
+**********************************************************************/
+
+#include "p2000t_mdcr.h"
+#include "emu.h"
+#include "formats/p2000t_cas.h"
```
To make the build behave consistently with and without prefix headers enabled, this needs to be:
```
#include "emu.h"
#include "p2000t_mdcr.h"
```
(i.e. prefix first, then module header)

pokowaka replied:
Will fix

cuavas commented on src/mame/machine/p2000t_mdcr.h:
```diff
@@ -0,0 +1,192 @@
+/// This particular phase decoder expects a signal of
+/// 1010 1010 which is used to derive the clock T.
+/// after a reset.
+class phase_decoder
```
This class name is too generic for the global namespace – please move this class inside the `mdcr_device` class or give it a less generic name.

pokowaka replied:
will fix

cuavas commented on src/mame/machine/p2000t_mdcr.h:
```diff
@@ -0,0 +1,192 @@
+	bool m_last_signal;
+	double m_tolerance;
+	int m_needs_sync;
+	timing_queue m_bit_queue;
```
How are you going to save/restore this variable length sequence for save state support?

pokowaka replied:
I was not aware of such a thing. Where do I find out more about save state support.

pokowaka commented:
Will fix whitespace and other issues. If possible please provide some guidance (examples) for state change support.

cuavas commented:
To support save states, you register objects to save/restore (usually in `device_start`), and optionally provide functions to call before saving or after loading memory (by overriding `device_pre_save` and/or `device_post_load`).  The objects to save need to be fixed size.  The core save/load code handles endianness (for well-beahved devices).  The key bindings to save and load states are Shift-F7/F7 by default.

Here’s a device that registers members for save/load: https://github.com/mamedev/mame/blob/mame0224/src/devices/sound/okim9810.cpp#L112

All the members registered using `save_item` will be saved/restored by the core save state functionality.  The `NAME` and `STRUCT_MEMBER` macros provide glue for saving simple members/arrays and members of structures in arrays, respectively.

This device needs to patch up some additional stuff after a save state is loaded, so it implements `device_post_load`: https://github.com/mamedev/mame/blob/mame0224/src/devices/sound/okim9810.cpp#L184

When the user loads a save state, this member function is called after all the items registered with `save_item` are restored.

pokowaka commented:
I had to use pre & post callbacks to store/recover tape postion, let me know if that is the right approach.

Tafoid commented:
The clone in the driver is throwing and exception at start
p2000m - p2000t.cpp
Exception at EIP=00000000034863f2 (mdcr_device::wda(int)+0x0002): ACCESS VIOLATION

pokowaka commented:
Investigating.
